### DISCUSSION TITLE
set_from_xresources failure

### DISCUSSION THREAD

**Discussion by anarcat** [i3/i3#5051] [Q&A] [answered]
Hi,

I'm trying to use set_from_resource in my i3 config and it doesn't seem to work: it always hits the fallback.

I have tried, for example:

    set_from_resource $green i3wm.color2 #f71000

i have this in my .Xresources:

    *.color2:       #519f50

... and the resulting "green" is "red" (#f710000 is red).

I am using py3status so maybe that could be the root of the problem? Yet the colors set in my i3 config *do* affect py3status, so I don't think the issue is there...

With debug logging, I do see this in the logs:

```
jui 12 14:25:14 angela i3[68836]: 2022-07-12 14:25:14 - ../src/config_parser.c:parse_file:992 - Could not get resource 'i3wm.color0', using fallback '#1C1B19'.
```

yet with `appres`, i do see that resource:

```
$ appres i3wm | grep color0:
*color0:        #1c1b19
```

this is i3 4.19 on Debian bullseye 11 (stable).

**Comment by Airblader** [comment]
You need to make sure that your X resources are loaded into the server _before_ i3 starts, the issue is likely that this currently happens _after_ i3 has started already. How you do that depends on how you start i3.

**Reply by anarcat** [reply]
i am pretty confident the resources are being loaded before i3 is being executed. but even if they weren't, i'm explicitly loading them after, then restarting i3: this should be enough for i3 to pick up the resources changes, as i3 re-executes itself... (or does i3 not reload resources on exec? that would be very weird)

**Reply by Airblader** [reply]
Can you try using `*color2` without the extra period? It's been a few years since I wrote xcb-util-xrm, I don't remember all the rules anymore.

**Reply by anarcat** [reply]
wow, okay, that works. WTH is going on here. `*.color2` should work, why doesn't this work?

**Reply by anarcat** [reply]
i think i understand now, `*.foo` will match only something like `App.Bar.foo`, and not `i3wm.foo`, because `*` and `.` are *two* bindings, not a single one. in other words, `*` is not a wildcard. I had to (re?)read the X(7) manpage to remember that. confusing as hell! i'll file a patch on the srcery theme, but it seems this is out of scope for i3, thanks!

**Reply by anarcat** [reply]
> I had to (re?)read the X(7) manpage to remember that. confusing as hell!

and now that i did re-re-read (for a third time!) it seems like it *should* have worked, but that's besides the point: it doesn't, so [hyrum's law](https://www.hyrumslaw.com/) and everything. :)

**Reply by Airblader** [reply]
I'll admit that reading the spec I agree with you right now, but I can also tell you that this is tested for in xcb-util-xrm, and those tests compare against the result of Xlib, so we're probably both missing something here. This spec is so damn confusing. I'll read this again tomorrow.

**Reply by Airblader** [reply]
What version of xcb-util-xrm is installed on your system? It's Debian stable, so probably ancient, which means this might just be https://github.com/Airblader/xcb-util-xrm/issues/79 and I just spent an hour looking at code that has been fixed four years ago. 😬😅

**Reply by anarcat** [reply]
uh, indeed, oops, that's my bad then. our version is perilously outdated:

https://tracker.debian.org/pkg/xcb-util-xrm

1.0! i'll file a bug in debian

update: i filed  https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1015738

**Comment by anarcat** [comment, marked as answer]
it seems i can't mark a specific reply as a solution, otherwise I would have marked https://github.com/i3/i3/discussions/5051#discussioncomment-3182474 as the solution. but basically, the problem is i was using `*.colorX` instead of plain `*colorX`.

This is an issue in the srcery theme, and I have proposed a fix for this in https://github.com/srcery-colors/srcery-terminal/pull/164.
